File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/decorator/solution/Barista.java
// Barista.java

import java.util.Vector;

public class Barista
{
	public Beverage makeBeverage(Vector<String> aDescription)
	{
		if (aDescription.size() == 0)
		{
			// throw an exception.
		}

		if (!aDescription.elementAt(0).equals("coffee"))
		{
			// throw an exception.
		}

		Beverage aBeverage = new Coffee();

		for (int i = 1; i < aDescription.size(); ++i)
		{
			if (aDescription.elementAt(i).equals("milk"))
			{
				aBeverage = new WithMilk(aBeverage);
			}
			else if (aDescription.elementAt(i).equals("sugar"))
			{
				aBeverage = new WithSugar(aBeverage);
			}
			else
			{
				// throw an exception.
			}
		}

		return aBeverage;
	}
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/decorator/solution/barista.cpp
// barista.cpp

#include <string>
#include <vector>
#include "barista.h"
#include "beverage.h"
#include "coffee.h"
#include "with_milk.h"
#include "with_sugar.h"

beverage * barista::make_beverage(const std::vector<std::string> & a_description)
{
	if (a_description.size() == 0)
	{
		// throw an exception.
	}

	if (a_description[0] != "coffee")
	{
		// throw an exception.
	}

	beverage * a_beverage = new coffee();

	for (std::vector<std::string>::size_type i = 1; i < a_description.size(); ++i)
	{
		if (a_description[i] == "milk")
		{
			a_beverage = new with_milk(a_beverage);
		}
		else if (a_description[i] == "sugar")
		{
			a_beverage = new with_sugar(a_beverage);
		}
		else
		{
			// throw an exception.
		}
	}

	return a_beverage;
}
